File: qualifiers/challenges/ropably/challenge/generator/checker.py
#!/usr/bin/env python3

import random
import string
import hmac
from typing import TextIO

function_prologue = '''
#include <stdio.h>
#include <stdlib.h>

int checker(char *input) {
    int result = 1;
'''

function_epilogue = '''
    return result;
}
'''

ALPHABET = string.ascii_letters + string.digits

def generate_checker(seed: bytes, out: TextIO) -> str:
    checker_seed = hmac.digest(seed, b'checker', 'sha256')
    random.seed(checker_seed)

    password = ''.join(random.choice(ALPHABET) for _ in range(16))
    password_values = password.encode()
    constraint_indices = list(range(len(password)))
    random.shuffle(constraint_indices)

    out.write(function_prologue + '\n')
    out.write(f'// password = "{password}"\n')

    for i in range(len(password)):
        idx_a = constraint_indices[(i+0) % len(constraint_indices)]
        idx_b = constraint_indices[(i+1) % len(constraint_indices)]
        a = password_values[idx_a]
        b = password_values[idx_b]

        constraint = 'result &= '
        match op := random.choice(['ADD', 'SUB', 'MUL', 'XOR']):
            case 'ADD':
                constant = (a + b) & 0xFFFF
                constraint += f'((input[{idx_a}] + input[{idx_b}]) & 0xFFFF) == {constant}'
            case 'SUB':
                constant = (a - b) & 0xFFFF
                constraint += f'((input[{idx_a}] - input[{idx_b}]) & 0xFFFF) == {constant}'
            case 'MUL':
                constant = (a * b) & 0xFFFF
                constraint += f'((input[{idx_a}] *  input[{idx_b}]) & 0xFFFF) == {constant}'
            case 'XOR':
                constant = (a ^ b) & 0xFFFF
                constraint += f'((input[{idx_a}] ^ input[{idx_b}]) & 0xFFFF) == {constant}'
            case _:
                raise ValueError(f'Unexpected operation: {op}')

        constraint += ';\n'
        out.write(constraint)
    out.write(function_epilogue + '\n')

    return password

File: qualifiers/challenges/ropably/attempt2/checker.c

#include <stdio.h>
#include <stdlib.h>

int checker(char *input) {
    int result = 1;

// password = "PpBOW2quokEiO2to"
result &= ((input[5] + input[13]) & 0xFFFF) == 100;
result &= ((input[13] + input[2]) & 0xFFFF) == 116;
result &= ((input[2] + input[8]) & 0xFFFF) == 177;
result &= ((input[8] ^ input[3]) & 0xFFFF) == 32;
result &= ((input[3] + input[4]) & 0xFFFF) == 166;
result &= ((input[4] + input[12]) & 0xFFFF) == 166;
result &= ((input[12] *  input[15]) & 0xFFFF) == 8769;
result &= ((input[15] + input[1]) & 0xFFFF) == 223;
result &= ((input[1] - input[9]) & 0xFFFF) == 5;
result &= ((input[9] - input[14]) & 0xFFFF) == 65527;
result &= ((input[14] *  input[7]) & 0xFFFF) == 13572;
result &= ((input[7] ^ input[0]) & 0xFFFF) == 37;
result &= ((input[0] *  input[11]) & 0xFFFF) == 8400;
result &= ((input[11] - input[6]) & 0xFFFF) == 65528;
result &= ((input[6] *  input[10]) & 0xFFFF) == 7797;
result &= ((input[10] + input[5]) & 0xFFFF) == 119;

    return result;
}

